File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/telemetry/analytics/MockAnalyticsCollector.kt
package stasis.test.client_android.lib.telemetry.analytics

import stasis.client_android.lib.telemetry.ApplicationInformation
import stasis.client_android.lib.telemetry.analytics.AnalyticsCollector
import stasis.client_android.lib.telemetry.analytics.AnalyticsEntry
import stasis.client_android.lib.telemetry.analytics.AnalyticsPersistence
import stasis.client_android.lib.utils.Try
import java.util.concurrent.atomic.AtomicReference

class MockAnalyticsCollector : AnalyticsCollector {
    private val entryRef: AtomicReference<AnalyticsEntry.Collected> =
        AtomicReference(AnalyticsEntry.collected(app = ApplicationInformation.none()))

    override fun recordEvent(name: String, attributes: Map<String, String>) {
        entryRef.updateAndGet { it.withEvent(name, attributes) }
    }

    override fun recordFailure(message: String) {
        entryRef.updateAndGet { it.withFailure(message) }
    }

    override fun state(): Try<AnalyticsEntry> =
        Try.Success(entryRef.get())

    override fun send() = Unit

    override val persistence: AnalyticsPersistence? =
        null
}

File: layers/src/test/scala/stasis/layers/telemetry/analytics/MockAnalyticsCollector.scala
package stasis.layers.telemetry.analytics

import java.util.concurrent.atomic.AtomicReference

import scala.concurrent.Future

import stasis.layers.telemetry.ApplicationInformation

class MockAnalyticsCollector extends AnalyticsCollector {
  private val mockPersistence = MockAnalyticsPersistence()
  private val entryRef: AtomicReference[AnalyticsEntry.Collected] =
    new AtomicReference(AnalyticsEntry.collected(app = ApplicationInformation.none))

  override def recordEvent(name: String, attributes: Map[String, String]): Unit = {
    val _ = entryRef.updateAndGet((entry: AnalyticsEntry.Collected) => entry.withEvent(name, attributes))
  }

  override def recordFailure(message: String): Unit = {
    val _ = entryRef.updateAndGet((entry: AnalyticsEntry.Collected) => entry.withFailure(message))
  }

  override def state: Future[AnalyticsEntry] =
    Future.successful(entryRef.get())

  override def send(): Unit = {
    val _ = mockPersistence.transmit(entryRef.get())
  }

  override def persistence: Option[AnalyticsPersistence] =
    Some(mockPersistence)
}
